### DISCUSSION TITLE
Modular implementation of two or more components

### DISCUSSION THREAD

**Discussion by H4ppy-04** [bevyengine/bevy#11145] [Q&A] [answered]
Reading through the [breakout example](https://github.com/bevyengine/bevy/blob/latest/examples/games/breakout.rs) I've noticed that *single functions* have been chosen to handle the functionality of all components[^1].

* Collision for multiple sprites are all handled in the [`check_for_collisions`](https://github.com/bevyengine/bevy/blob/22e39c4abf6e2fdf99ba0820b3c35db73be71347/examples/games/breakout.rs#L345-L398) function. 
* The [`apply_velocity`](https://github.com/bevyengine/bevy/blob/22e39c4abf6e2fdf99ba0820b3c35db73be71347/examples/games/breakout.rs#L333C1-L338C2) function manages the speed of multiple components.

**This begs the question**: *What if components have individual functionality such as differing speed and/or acceleration?* Movement and any other attributes that apply to specific components would have to be handled on an individual level.

Given components with individual properties:

```rust
#[derive(Component)]
struct Projectile {
    position: Vec2,
    velocity: Vec2
}

#[derive(Component)]
struct Enemy {
    health: i32
}
```

How might one go about using `impl` to implement *individual functionality* for `Projectile` and `Enemy`? Thank you in advance.

[^1]: This has [been proven to be the most efficient approach](https://ieeexplore.ieee.org/document/5298282) in this context.

**Comment by ThierryBerger** [comment, marked as answer]
You have different options:

- You can add more data, so you can fine tune each entity.
```rs
/// The current acceleration, so the speed can augment progressively
#[derive(Component)]
struct Acceleration {
    current: f32,
}
/// a configuration for each entity, like the acceleration rate.
#[derive(Component)]
struct AccelerationChange {
    change_per_frame: f32,
}
/// Another configuration to know when the acceleration is too much, or not get an unsupported speed.
#[derive(Component)]
struct MaxSpeed(f32);
```

- Those additional data could also be stored in a `Resource`, might be interesting if you have multiple similarly configured entities.
- If you need more specific logic, it might be practical to "specialize" each system, by adding marker components:

```rs

#[derive(Component)]
struct MarkerFastProjectile;
#[derive(Component)]
struct MarkerSinProjectile;

fn system_move_fast(mut q_projectiles: Query<&mut Transform, With<MarkerFastProjectile>>) {
  for t in q_projectiles.iter() {
    // We could modify Speed, to defer actual movement to another system
    t.translation += Vec3::X * 5f32;
  }
}

fn system_move_sin(time: Res<Time>, mut q_projectiles: Query<&mut Transform, With<MarkerSinProjectile>>) {
  for t in q_projectiles.iter() {
    t.translation += Vec3::X * 1f32 + Vec3::Y * time.elapsed_seconds().sin();
  }
}

```
